src/Middleware/ThrottleMiddleware.php

```diff
@@ -45,11 +45,25 @@ public function __invoke(ServerRequestInterface $request, ResponseInterface $res
         $this->_initCache();
         $this->_count = $this->_touch();
 
-        if ($this->_count > $this->getConfig('limit')) {
+        $config = $this->getConfig();
+
+        if ($this->_count > $config['limit']) {
             $stream = new Stream('php://memory', 'wb+');
-            $stream->write((string)$this->getConfig('message'));
+
+            if (is_array($config['response']['headers'])) {
+                foreach ($config['response']['headers'] as $name => $value) {
+                    $response = $response->withHeader($name, $value);
+                }
+            }
+
+            if (isset($config['message'])) {
+                $stream->write((string)$config['message']);
+            } else {
+                $stream->write((string)$config['response']['body']);
+            }
 
             return $response->withStatus(429)
+                ->withType($config['response']['type'])
```

Comment:
`withType()` is not a method of `ResponseInterface` so it would be better to avoid using it. Also using this method makes [phpstan]() unhappy as the `$response` argument is typehinted as `Psr\Http\Message\ResponseInterface` not `Cake\Http\Response`. 

Using a separate `type` key inside `response` isn't really necessary anyway. One can just set `Content-Type` in `headers`.

Reply:
Well as I've explained sooner today, when we set the stream with the `withBody` method method, Cake overwrites `Content-Type` header... That's why I've set back the `type` handling

Reply:
Okay, then let's just remove the typehints for arguments of `__invoke()` method. Since the method isn't going to be called directly by user the typehints aren't really useful anyway. Also update the docs for request and response argument to use `Cake\Network\Request|Reponse` instead of the interfaces.

P. S. Don't forget to remove the "use" statements for the interfaces at top of file :)

Reply:
I may have a solution, lemme 5 minutes

Reply:
Yeah this middleware isn't really framework agnostic so trying to stick to the PSR7 interfaces doesn't help.

Reply:
I've tried something (sorry for the whole process, first time with PSR7 stuff :grimacing:)